pkg/datapath/iptables/iptables.go

```diff
@@ -1036,6 +1036,17 @@ func (m *IptablesManager) InstallRules(ifName string) error {
 		}
 	}
 
+	// EKS expects the aws daemonset to set up specific rules for marking
+	// packets It marks packets coming from another node and restores the
+	// mark on the return path to force a lookup into the main routing
+	// table. We want to reproduce something similar. Please see note in
+	// Reanitialize() in package "loader" for more details.
+	if option.Config.Masquerade && option.Config.IPAM == ipamOption.IPAMENI {
```

Comment:
I looked at it again, and making it depend on masquerading seems wrong.

If masquerade is disabled here, there is no visible impact on most of the connectivity checks (in particular, the one we focus on is still working). Only the ones related to external connectivity fail to get ready. I think my initial idea was “we always need masquerading with EKS to have external connectivity, so I can rely on it to pin down the « EKS » case”, but on second thought this seems dubious. And there are probably no particular reason why I should apply the rules on the masqueraded interfaces either.

So there are two things that I need to sort out here:
- Should I restrict those rules to EKS only, and if so, how can I do it? At least it's easy to restrict to AWS. Ideally we _should_ restrict further to EKS (as far as I know, this fix is not required for regular AWS), but do we have an easy way to do so (and is it worth it)?
- How should I determine the interface on which the rules should be applied? For the GSG it's easy, it's just `eth0` but I can't really go with that in the code. So I need a way to get the list of interfaces through which the node is susceptible to receive traffic from other nodes from the cluster, do we have a way to retrieve that? Other iptables/ip rules I found were only on `lxc*`, `cilium_*` interfaces.

Reply:
> Should I restrict those rules to EKS only, and if so, how can I do it? At least it's easy to restrict to AWS. Ideally we should restrict further to EKS (as far as I know, this fix is not required for regular AWS), but do we have an easy way to do so (and is it worth it)?

I think that restricting using `IPAMENI` is very reasonable and handles the case we want. Strictly speaking you might interpret that as handling only IPAM via ENI (rather than routing/devices), but in practice these two are tied together. If we're in an environment where we're handling direct IPAM via ENI, then there are potentially multiple ENI devices that we need to manage, and that's the case where this comes up.

> How should I determine the interface on which the rules should be applied? For the GSG it's easy, it's just eth0 but I can't really go with that in the code. So I need a way to get the list of interfaces through which the node is susceptible to receive traffic from other nodes from the cluster, do we have a way to retrieve that? Other iptables/ip rules I found were only on lxc*, cilium_* interfaces.

On latest code, does `--devices` provide what you need? That's where we detect devices with access outside the cluster AFAIK (including on v1.8.x). @pchaigno / @brb can keep me honest here. But then we'll still need to figure it out for older releases as well. On older release there was `--device` but I don't know whether ENI users will be guaranteed to configure it. May be worth cross-referencing with the EKS guide.

Reply:
> How should I determine the interface on which the rules should be applied?

If the device is used to communicate between k8s nodes, then you can use `DirectRoutingDevice`.

Reply:
> Is this the best location to add the rules?

Just thinking loud but we have another PR which adds custom iptables rules for GKE to nodeinit: https://github.com/cilium/cilium/pull/11782/files. Perhaps we should not spread cloud provider specific rules across multiple places, and have a dedicated package for it?

Reply:
The GKE bits are likely to be removed again fairly soon given that native integration.

Reply:
Main topic of the discussion addressed in the sense that the added rules do not depend on masquerading anymore.

The other PR has been merged already and according to Thomas, it may not stay long in the code, so I won't address the suggestion from Martynas on creating a dedicated package for the rules. Could be done as a follow-up, if desired and relevant.